File: insight_desktop/engine/prompt_builder.py
"""Assembles the exact message list sent to the LLM for a single turn."""

from __future__ import annotations

from engine.visual_context import VisualContext


class PromptBuilder:
    def build(
        self,
        personality_prompt: str,
        memory_facts: list[str],
        history_messages: list[dict],
        history_summary_note: str | None,
        current_utterance: str,
        visual_context: VisualContext | None = None,
    ) -> tuple[list[dict], str]:
        system_content = personality_prompt.strip()

        if memory_facts:
            facts_block = "\n".join(f"- {fact}" for fact in memory_facts)
            system_content += f"\n\nThings you know about the user (long-term memory):\n{facts_block}"

        if visual_context is not None:
            system_content += f"\n\n{visual_context.prompt_block()}"

        if history_summary_note:
            system_content += f"\n\n{history_summary_note}"

        messages: list[dict] = [{"role": "system", "content": system_content}]
        messages.extend(history_messages)
        messages.append({"role": "user", "content": current_utterance})

        debug_text = "\n\n".join(f"[{m['role'].upper()}]\n{m['content']}" for m in messages)
        return messages, debug_text

File: insight_ios/Packages/InsightCore/Sources/InsightCore/PromptBuilder.swift
import Foundation

/// Assembles the exact message list sent to the LLM for a single turn.
public struct PromptBuilder: Sendable {
    public init() {}

    public func build(
        personalityPrompt: String,
        memoryFacts: [String],
        historyMessages: [ChatMessage],
        historySummaryNote: String?,
        currentUtterance: String,
        visualContext: VisualContext? = nil
    ) -> (messages: [ChatMessage], debugText: String) {
        var systemContent = personalityPrompt.trimmingCharacters(in: .whitespacesAndNewlines)

        if !memoryFacts.isEmpty {
            let factsBlock = memoryFacts.map { "- \($0)" }.joined(separator: "\n")
            systemContent += "\n\nThings you know about the user (long-term memory):\n\(factsBlock)"
        }

        if let visualContext {
            systemContent += "\n\n\(visualContext.promptBlock())"
        }

        if let historySummaryNote {
            systemContent += "\n\n\(historySummaryNote)"
        }

        var messages: [ChatMessage] = [ChatMessage(role: "system", content: systemContent)]
        messages.append(contentsOf: historyMessages)
        messages.append(ChatMessage(role: "user", content: currentUtterance))

        let debugText = messages
            .map { "[\($0.role.uppercased())]\n\($0.content)" }
            .joined(separator: "\n\n")

        return (messages, debugText)
    }
}
